src/webgpu/util/floating_point.ts

```diff
@@ -708,6 +709,41 @@ export abstract class FPTraits {
     return new FPInterval(this.kind, n, n);
   }
 
+  /**
+   * WGSL specifies unbounded precision:
+   *   https://www.w3.org/TR/WGSL/#floating-point-accuracy
+   * In most computations doubles (number in js) are equivalent
+   * to correctly rounded intervals. However in some cases
+   * (addition/subtraction) doubles do not provide enough numeric precision.
+   *
+   * These cases are whenever a number, x (e.g. 1.0) is added (or subtracted)
+   * from a much smaller number, y (e.g. smallest positive normal), such that
+   * the difference between them is smaller then ULP(x). When working in JS
+   * numbers the result will, incorrectly, simply be x, since it will get
+   * rounded. We must detect these cases where double precision does not
```

Comment:
I think this also has to say that WGSL allows rounding either up or down, so it may not compute the same rounded result as in JS.

JS uses round-ties-to-even.  https://tc39.es/ecma262/#sec-ecmascript-language-types-number-type